### PR TITLE
MW-1031: getQuestionQttributesForEMPerformance

### PR DESCRIPTION
**Pull request by lajosarpad**
This is a performance optimization that affects survey settings tab loading.

Measurements:

Text elements	6.98s	0.49s
Privacy policy	 	6.93s	0.65s
Presentation	 	6.8s	0.45s
Participant settings	 	6.95s	0.45s
Notifications & data	 	6.87s	0.43s
Publication & access	 	6.79s	0.46s
Panel integration	 	8.23s	1.12s
Resources	 	6.89s	0.44s
Simple plugins	 	10.67s	0.46s

### CODE REVIEW

**Review by c-schmitz — CHANGES_REQUESTED** [review]
Code looks fine, but something broke in the tests. Please check.

**Review by Shnoulle — APPROVED** [review]
Avoid call of oQuestion if already loaded :+1:

**Comment by Shnoulle** [comment]
About QuestionAttribute : a big difference between 3 and 4 are here too.

In 3 QuestionAttributedefinition was static, it's not the case (even for core part) in 4.

See [18475: Multiple event newQuestionAttributes happen](https://bugs.limesurvey.org/view.php?id=18475)

I think a good improvement can create core + plugin static function/var + update it with theme after ?

File: application/models/QuestionAttribute.php [hunk 1]
```diff
@@ -219,16 +219,21 @@ public function setMultiple($iSid, $aQids, $aAttributesToUpdate, $aValidQuestion
...
      * @access public
-     * @param int $iQuestionID
+     * @param int|Question $q
      * @param string $sLanguage restrict to this language (if null $oQuestion->survey->allLanguages will be used)
      * @return array|false
      *
      * @throws CException throws exception if questiontype is null
      */
-    public function getQuestionAttributes($iQuestionID, $sLanguage = null)
+    public function getQuestionAttributes($q, $sLanguage = null)
```
Comment (olleharstedt) [inline application/models/QuestionAttribute.php:228]: It wasn't possible to change all call locations to this method to always only pass object as input?

Comment (lajosarpad) [reply application/models/QuestionAttribute.php:228]: We are currently working on performance issues. We could refactor the further references, but that would require a separate ticket and an approval

image

We can only change the parameter to accept question objects if we change and test all references. That would take a while and in the meantime the users would not get the benefits of our results. If we can create a ticket to refactor all references and then the parameter of this method, that would be great, I agree with you. But even then we should break it up to smaller PRs with the last one being the change of the parameter to not allow integers.

Comment (olleharstedt) [reply application/models/QuestionAttribute.php:228]: Wow that's a lot of call sites.. Got it. :sweat_smile:

**Comment by lajosarpad** [comment]
> Code looks fine, but something broke in the tests. Please check.

All green now

File: application/helpers/expressions/em_manager_helper.php [hunk 2]
```diff
@@ -8275,36 +8275,27 @@ private function getQuestionAttributesForEM($surveyid = 0, $qid = 0, $lang = '')
         if ($qid) {
             $oQids = Question::model()->findAll(
                 [
-                    'select'    => 'qid',
-                    'group'     => 'qid',
-                    'distinct'  => true,
                     'condition' => "qid=:qid and parent_qid=0",
```
Comment (Shnoulle) [inline application/helpers/expressions/em_manager_helper.php:8278]: We better understand why the name it's $oQids and not $oQuestion : i think it's a good idea to rename this variable here.

**Review by Shnoulle — COMMENTED** [review]
Maybe rename variable oQids to oQuestion ?

**Review by c-schmitz — APPROVED** [review]

File: application/helpers/expressions/em_manager_helper.php (hunk 2 above)
Comment (Shnoulle) [reply application/helpers/expressions/em_manager_helper.php:8278]: Thank you

**Comment by lajosarpad** [comment]
@Dennis1993 @c-schmitz the change was implemented accordingly

**Comment by olleharstedt** [comment]
> Maybe rename variable oQids to oQuestion ?

Or get rid of Hungarian notation and just use `$question` :sweat_smile:

**Comment by Shnoulle** [comment]
> > Maybe rename variable oQids to oQuestion ?
> 
> Or get rid of Hungarian notation and just use `$questions` 😅

One or another, but no oQid ;)

**Pull request merged**